msm/src/precomputed_srs.rs

```diff
@@ -1,18 +1,145 @@
+    }
+    PathBuf::from(base_path).join("../srs/test_bn254.srs")
+}
+
+/// Tries to read the SRS from disk, otherwise panics. Returns the
+/// value without Lagrange basis.
+fn read_bn254_srs_from_disk(srs_path: PathBuf) -> PairingSRS<BN254> {
+    let file =
+        File::open(srs_path.clone()).unwrap_or_else(|_| panic!("missing SRS file: {srs_path:?}"));
+    let reader = BufReader::new(file);
+    let srs: TestPairingSRS<BN254> = rmp_serde::from_read(reader).unwrap();
+    From::from(srs)
+}
+
+/// Creates a BN254 SRS. If the `overwrite_srs` flag is on, or
+/// `SRS_OVERWRITE` env variable is ON, also writes it into the file.
+fn create_and_store_srs_with_path(
+    force_overwrite: bool,
+    domain_size: usize,
+    srs_path: PathBuf,
+) -> PairingSRS<BN254> {
+    // We generate with a fixed-seed RNG, only used for testing.
+    let mut rng = &mut StdRng::from_seed([42u8; 32]);
+    let trapdoor = Fp::rand(&mut rng);
+    let mut srs = PairingSRS::create(trapdoor, domain_size);
+
+    for sub_domain_size in 1..=domain_size {
+        let domain = EvaluationDomains::<Fp>::create(sub_domain_size).unwrap();
+        srs.full_srs.add_lagrange_basis(domain.d1);
+    }
+
+    // overwrite SRS if the env var is set
+    if force_overwrite || std::env::var("SRS_OVERWRITE").is_ok() {
+        // Create parent directories
+        std::fs::create_dir_all(srs_path.parent().unwrap()).unwrap();
+        // Open/create the file
+        let mut file = std::fs::OpenOptions::new()
+            .create(true)
+            .write(true)
+            .open(srs_path.clone())
+            .expect("failed to open SRS file");
+
+        let test_srs: TestPairingSRS<BN254> = From::from(srs.clone());
+        let srs_bytes = rmp_serde::to_vec(&test_srs).unwrap();
+        file.write_all(&srs_bytes).expect("failed to write file");
+        file.flush().expect("failed to flush file");
+    }
+
+    // get SRS from disk
+    let srs_on_disk = read_bn254_srs_from_disk(srs_path);
+
+    // check that it matches what we just generated
+    assert_eq!(srs, srs_on_disk);
+
     srs
 }
+
+/// Creates and writes the SRS into `get_bn254_srs_path()`.
+pub fn create_and_store_srs(force_overwrite: bool, domain_size: usize) -> PairingSRS<BN254> {
+    let srs_path = get_bn254_srs_path();
+    create_and_store_srs_with_path(force_overwrite, domain_size, srs_path)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    /// Creates or checks the real full-sized SRS. Only overwrites
+    /// with the ENV flag.
+    pub fn create_or_check_srs() {
+        let domain_size = DOMAIN_SIZE;
+        create_and_store_srs_with_path(false, domain_size, get_bn254_srs_path());
+    }
+
+    #[test]
+    /// Fast test for a small-sized SRS. Always writes & reads.
+    pub fn check_bn256_srs_serialization() {
+        let domain_size = 1 << 8;
+        let test_srs_path = PathBuf::from("/tmp/test_bn254.srs");
```

Comment:
I hope it works in the CI (and other platforms - permissions for instance). I would use https://doc.rust-lang.org/std/env/fn.temp_dir.html